File: apps/sdk/src/events/store.py
from typing import List, Protocol

from src.events.schemas import AgentKitEvent


class EventStore(Protocol):
    """Interface for a persisted event log."""

    async def append(self, event: AgentKitEvent) -> None:
        """Append an event to the store."""
        ...

    async def get_events(self, session_id: str) -> List[AgentKitEvent]:
        """Retrieve all events for a session, ordered by time."""
        ...


class InMemoryEventStore:
    """Simple in-memory event store for testing and simple deployments."""

    def __init__(self):
        self._events: dict[str, List[AgentKitEvent]] = {}

    async def append(self, event: AgentKitEvent) -> None:
        if event.session_id not in self._events:
            self._events[event.session_id] = []
        self._events[event.session_id].append(event)
        # Sort by timestamp just in case
        self._events[event.session_id].sort(key=lambda e: e.timestamp)

    async def get_events(self, session_id: str) -> List[AgentKitEvent]:
        return self._events.get(session_id, []).copy()

File: agentkit/ts/src/events/store.ts
/**
 * Event store interface and in-memory backend, mirroring
 * `agentkit.infra.events.store`. The append-only log is the source of truth;
 * session state is a projection (see `replaySession`).
 */
import type { AgentKitEvent } from "./schemas";

/** Interface every persistent backend implements. */
export interface EventStore {
  /** Append an event to the store. */
  append(event: AgentKitEvent): Promise<void>;
  /** Retrieve all events for a session, ordered by time. */
  getEvents(sessionId: string): Promise<AgentKitEvent[]>;
}

/**
 * In-memory event store for tests and simple deployments. Events live in a
 * per-session map and are lost on process exit.
 */
export class InMemoryEventStore implements EventStore {
  private readonly events = new Map<string, AgentKitEvent[]>();

  async append(event: AgentKitEvent): Promise<void> {
    const bucket = this.events.get(event.session_id) ?? [];
    bucket.push(event);
    bucket.sort((a, b) => a.timestamp - b.timestamp);
    this.events.set(event.session_id, bucket);
  }

  async getEvents(sessionId: string): Promise<AgentKitEvent[]> {
    return [...(this.events.get(sessionId) ?? [])];
  }
}
